Aufgabe 38.

Der Datensatz Zement enthält Festigkeitswerte (Y) für Zement von drei Zuliefern, der mit einem Additiv versetzt wurde. Zusätzlich wurde die Feuchtigkeit des Zements gemessen.

In [15]:
import pandas as pd
from utils import *
from statsmodels.api import OLS, add_constant
import seaborn as sns
import matplotlib.pyplot as plt

In [16]:
sns.set(style="whitegrid")
pd.set_option('display.max_columns', 100)
healthy_orange = '#EC6602'
healthy_orange_50 = '#F9B591'
healthy_orange_25 = '#FDDDCB'
siemens_petrol = '#009999'
siemens_petrol_50 = '#87D2D2'
sns.set_palette([healthy_orange, siemens_petrol, healthy_orange_50, healthy_orange_25, siemens_petrol_50])
RANDOM_SEED = 42

In [17]:
df = pd.read_csv('../data/zement.txt')
df.head()

Festigkeit    Marke     Additiv  Feuchtigkeit
0        32.6  Grauert  Verstärker          56.4
1        31.5  Grauert  Verstärker          44.1
2        30.0  Grauert  Verstärker          50.0
3        29.8  Grauert    Standard          52.5
4        29.7  Grauert  Verstärker          46.7

In [18]:
df = pd.get_dummies(df, columns=['Marke', 'Additiv'], dtype=int)
df.head()

Festigkeit  Feuchtigkeit  Marke_Grauert  Marke_Konsort  Marke_Zemix  \
0        32.6          56.4              1              0            0   
1        31.5          44.1              1              0            0   
2        30.0          50.0              1              0            0   
3        29.8          52.5              1              0            0   
4        29.7          46.7              1              0            0   

   Additiv_Standard  Additiv_Verstärker  
0                 0                   1  
1                 0                   1  
2                 0                   1  
3                 1                   0  
4                 0                   1

In [19]:
df = df.drop(columns=['Marke_Zemix', 'Additiv_Verstärker'])

## Base Modell

In [20]:
X = df.drop(columns=['Festigkeit'])
y = df['Festigkeit']
X = add_constant(X)
model = OLS(y, X).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             Festigkeit   R-squared:                       0.757
Model:                            OLS   Adj. R-squared:                  0.718
Method:                 Least Squares   F-statistic:                     19.48
Date:                Sun, 08 Jun 2025   Prob (F-statistic):           2.17e-07
Time:                        01:03:09   Log-Likelihood:                -57.794
No. Observations:                  30   AIC:                             125.6
Df Residuals:                      25   BIC:                             132.6
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const               41.6897      2.960  

a) Passen Sie ein lineares Regressionsmodell mit allen Haupteffekten, Wechselwirwirkungen und quadratischen Effekten an.

In [21]:
# Polynomiale Features
df['Feuchtigkeit_squared'] = df['Feuchtigkeit'] ** 2

# Wechselwirkungen
df['Konsort_x_Feuchtigkeit'] = df['Marke_Konsort'] * df['Feuchtigkeit']
df['Grauert_x_Feuchtigkeit'] = df['Marke_Grauert'] * df['Feuchtigkeit']
df['Standard_x_Feuchtigkeit'] = df['Additiv_Standard'] * df['Feuchtigkeit']
df['Konsort_x_Standard'] = df['Marke_Konsort'] * df['Additiv_Standard']
df['Grauert_x_Standard'] = df['Marke_Grauert'] * df['Additiv_Standard']
df.head()

Festigkeit  Feuchtigkeit  Marke_Grauert  Marke_Konsort  Additiv_Standard  \
0        32.6          56.4              1              0                 0   
1        31.5          44.1              1              0                 0   
2        30.0          50.0              1              0                 0   
3        29.8          52.5              1              0                 1   
4        29.7          46.7              1              0                 0   

   Feuchtigkeit_squared  Konsort_x_Feuchtigkeit  Grauert_x_Feuchtigkeit  \
0               3180.96                     0.0                    56.4   
1               1944.81                     0.0                    44.1   
2               2500.00                     0.0                    50.0   
3               2756.25                     0.0                    52.5   
4               2180.89                     0.0                    46.7   

   Standard_x_Feuchtigkeit  Konsort_x_Standard  Grauert_x_Standard  
0                      0.0                   0                   0  
1                      0.0                   0                   0  
2                      0.0                   0                   0  
3                     52.5                   0                   1  
4                      0.0                   0                   0

In [22]:
x = df.drop(columns=['Festigkeit'])
y = df['Festigkeit']
x = add_constant(x)
model = OLS(y, x).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             Festigkeit   R-squared:                       0.821
Model:                            OLS   Adj. R-squared:                  0.726
Method:                 Least Squares   F-statistic:                     8.687
Date:                Sun, 08 Jun 2025   Prob (F-statistic):           3.43e-05
Time:                        01:03:09   Log-Likelihood:                -53.256
No. Observations:                  30   AIC:                             128.5
Df Residuals:                      19   BIC:                             143.9
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                     

Ich nehme jetzt `Standard_x_Feuchtigkeit` raus, weil es P-Wert der großten P-Wert hat.

In [23]:
x = df.drop(columns=['Festigkeit', 'Standard_x_Feuchtigkeit'])
y = df['Festigkeit']
x = add_constant(x)
model = OLS(y, x).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             Festigkeit   R-squared:                       0.820
Model:                            OLS   Adj. R-squared:                  0.739
Method:                 Least Squares   F-statistic:                     10.14
Date:                Sun, 08 Jun 2025   Prob (F-statistic):           1.04e-05
Time:                        01:03:09   Log-Likelihood:                -53.284
No. Observations:                  30   AIC:                             126.6
Df Residuals:                      20   BIC:                             140.6
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     18

Jetzt nehmen ich `Konsort_x_Standard` raus.

In [24]:
x = df.drop(columns=['Festigkeit', 'Standard_x_Feuchtigkeit', 'Konsort_x_Standard'])
y = df['Festigkeit']
x = add_constant(x)
model = OLS(y, x).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             Festigkeit   R-squared:                       0.814
Model:                            OLS   Adj. R-squared:                  0.743
Method:                 Least Squares   F-statistic:                     11.48
Date:                Sun, 08 Jun 2025   Prob (F-statistic):           3.97e-06
Time:                        01:03:09   Log-Likelihood:                -53.797
No. Observations:                  30   AIC:                             125.6
Df Residuals:                      21   BIC:                             138.2
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     18

In [25]:
x = df.drop(columns=['Festigkeit', 'Standard_x_Feuchtigkeit', 'Konsort_x_Standard'])
y = df['Festigkeit']
x = add_constant(x)
model = OLS(y, x).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             Festigkeit   R-squared:                       0.814
Model:                            OLS   Adj. R-squared:                  0.743
Method:                 Least Squares   F-statistic:                     11.48
Date:                Sun, 08 Jun 2025   Prob (F-statistic):           3.97e-06
Time:                        01:03:09   Log-Likelihood:                -53.797
No. Observations:                  30   AIC:                             125.6
Df Residuals:                      21   BIC:                             138.2
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     18

Jetzt nehme ich `Konsort_x_Feuchtigkeit` raus.

In [26]:
x = df.drop(columns=['Festigkeit', 'Standard_x_Feuchtigkeit', 'Konsort_x_Standard',
                    'Konsort_x_Feuchtigkeit'])
y = df['Festigkeit']
x = add_constant(x)
model = OLS(y, x).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             Festigkeit   R-squared:                       0.807
Model:                            OLS   Adj. R-squared:                  0.746
Method:                 Least Squares   F-statistic:                     13.15
Date:                Sun, 08 Jun 2025   Prob (F-statistic):           1.47e-06
Time:                        01:03:09   Log-Likelihood:                -54.337
No. Observations:                  30   AIC:                             124.7
Df Residuals:                      22   BIC:                             135.9
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     20

Jetzt nehme ich `Feuchtigkeit_squared` raus.

In [27]:
x = df.drop(columns=['Festigkeit', 'Standard_x_Feuchtigkeit', 'Konsort_x_Standard',
                    'Konsort_x_Feuchtigkeit', 'Feuchtigkeit_squared'])
y = df['Festigkeit']
x = add_constant(x)
model = OLS(y, x).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             Festigkeit   R-squared:                       0.794
Model:                            OLS   Adj. R-squared:                  0.741
Method:                 Least Squares   F-statistic:                     14.80
Date:                Sun, 08 Jun 2025   Prob (F-statistic):           7.05e-07
Time:                        01:03:09   Log-Likelihood:                -55.307
No. Observations:                  30   AIC:                             124.6
Df Residuals:                      23   BIC:                             134.4
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     44

!!R2 und R2-Adjusted die beide sind jetzt schlechter, deswegen ich würde `Feuchtigkeit_squared` drin lassen.

In [28]:
x = df.drop(columns=['Festigkeit', 'Standard_x_Feuchtigkeit', 'Konsort_x_Standard', 'Additiv_Standard',
                    'Konsort_x_Feuchtigkeit'])
y = df['Festigkeit']
x = add_constant(x)
model = OLS(y, x).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             Festigkeit   R-squared:                       0.795
Model:                            OLS   Adj. R-squared:                  0.742
Method:                 Least Squares   F-statistic:                     14.90
Date:                Sun, 08 Jun 2025   Prob (F-statistic):           6.63e-07
Time:                        01:03:09   Log-Likelihood:                -55.225
No. Observations:                  30   AIC:                             124.5
Df Residuals:                      23   BIC:                             134.3
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     17

Das ist mein finales Modell.

b) Reduzieren Sie das Modell in geeigneter Weise. Wie haben sich das Bestimmtheitsmaß und das korrigierte Bestimmtheitsmaß im Vergleich zum vollen Modell verändert?

- Das Modell mit allen Variablen hatte R2 von 0.821 und R2-Adjusted von 0.726.
- Das Modell am Ende nach der Entfernung nicht signifikanten Variablen bzw. (wo wir keine große Verlust an R2 hatten) haben wir R2 von 0.807 und R2-Adjusted von 0.746.
- Unser Base Modell hatte den R2 von 0.757 und R2-Adjusted von 0.718.

c) Wieso werden nicht-signifikante Effekte aus einem Regressionsmodell entfernt?

Um R2-Adjusted zu verbessern bzw. unser Modell robuster zu machen.

d) Was unterscheidet den Faktor Feuchtigkeit von den anderen beiden Faktoren? (Hier ist nicht die Skala gemeint)

### Frage: 
Wir haben immer k-1 Kategorien genommen und daraus Interaktionen gebildet, hätten wir doch all k Kategorien für Interaktion genommen bzw. eine andere Kategorie rausgelassen und dann Interaktion gebildet, hätte es einen Unterschied gemacht? Ich habe das schonmal getestet und mind. R2 ist immer das Gleiche geblieben, aber das kann nicht nachvollziehen warum es keinen Unterschied macht? Kann es schon sein dass eine Kategorie zusammen mit einem anderen Prädiktor gute Korrelation mit der Zielvariable hat als eine andere Kategorie mit dem gleichen Prädiktor, aber trotzdem macht es keinen Unterschied.